confluent_kafka/avro/serializer/__init__.py

```diff
@@ -33,8 +61,270 @@ def __str__(self):
+
+        subject = self.subject_strategy(topic, self.is_key, get_schema(record))
+
+        schema_id = self.registry_client.register(subject, get_schema(record))
+        if not schema_id:
+            message = "Failed to register schema with subject {}.".format(subject)
+            raise SerializerError(message, is_key=self.is_key)
+
+        if schema_id not in self.codec_cache:
+            self.codec_cache[schema_id] = self._get_encoder_func(get_schema(record))
+
+        return self._encode(schema_id, record)
+
+    def _get_encoder_func(self, writer_schema):
+        if HAS_FAST:
+            return lambda record, fp: schemaless_writer(fp, writer_schema.to_json(), record)
+        writer = avro.io.DatumWriter(writer_schema)
+        return lambda record, fp: writer.write(record, avro.io.BinaryEncoder(fp))
+
+    def _encode(self, schema_id, datum):
+        """
+        Encode a datum with a given schema id.
+        :param int schema_id: integer ID
+        :param object datum: An object to serialize
+        :param bool is_key: If the record is a key
+        :param SerializerErr err_type: Error type to raise on serialization exception
+        :returns: decoder function
+        :rtype: func
+        """
+
+        if schema_id not in self.codec_cache:
+            try:
+                schema = self.registry_client.get_by_id(schema_id)
+                if not schema:
+                    raise SerializerError("Schema does not exist", self.is_key)
+                self.codec_cache[schema_id] = self._get_encoder_func(schema)
+            except ClientError:
+                exc_type, exc_value, exc_traceback = sys.exc_info()
+                raise SerializerError(
+                    repr(traceback.format_exception(exc_type, exc_value, exc_traceback)),
+                    self.is_key)
+
+        # get the writer
+        writer = self.codec_cache[schema_id]
+        with ContextStringIO() as outf:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            outf.write(struct.pack('>bI', MAGIC_BYTE, schema_id))
+
+            # write the record to the rest of the buffer
+            writer(datum, outf)
+            return outf.getvalue()
+
+
+class AvroDeserializer(object):
+    """
+    Decodes Kafka messages encoded by Confluent Schema Registry compliant Avro Serializers.
+
+    :param registry_client CachedSchemaRegistryClient: Instance of CachedSchemaRegistryClient.
+    :param bool is_key: True if configured as a key_serializer.
+    :param schema reader_schema: Optional reader schema to be used during deserialization.
+    """
+    __slots__ = ["registry_client", "codec_cache", "is_key", "reader_schema"]
+
+    def __init__(self, registry_client, is_key=False, reader_schema=None):
+        self.registry_client = registry_client
+        self.codec_cache = {}
+        self.is_key = is_key
+        self.reader_schema = reader_schema
+
+    def __call__(self, topic, datum):
```

Comment:
`datum` is not a good name for the serialized data, as it suggests the actual object itself (especially in avro terminology). maybe `payload`?